File: TTT3D/player.cpp
#include "player.hpp"
#include <cstdlib>

#define NaN 0x7fffffff
#define DEPTH 2

namespace TICTACTOE3D {

    GameState Player::play(const GameState &pState, const Deadline &pDue) {


        vector<GameState> lNextStates;
        pState.findPossibleMoves(lNextStates);

        if (lNextStates.size() == 0) {
            return GameState(pState, Move());
        }

        /*
        * Here you should write your clever algorithms to get the best next move, ie the best
        * next state. This skeleton returns a random move instead.
        */

//      return lNextStates[rand() % lNextStates.size()];

        currentPlayer = pState.getNextPlayer();
        uint8_t nextPlayer;
        if (currentPlayer == CELL_X) {
            nextPlayer = CELL_O;
        } else {
            nextPlayer = CELL_X;
        }

        int maxHeuristicVal = -NaN;
        int maxValIndex = 0;
        int heuristicVal;
        int alpha = -NaN;
        int beta = NaN;

        for (int i = 0; i < lNextStates.size(); i++) {
            heuristicVal = alphaBeta(lNextStates[i], nextPlayer, DEPTH, alpha, beta);
            if (heuristicVal > maxHeuristicVal) {
                maxHeuristicVal = heuristicVal;
                maxValIndex = i;
            }
        }
        return lNextStates[maxValIndex];
    }


    int Player::getHeuristicVal(const GameState &pState, const uint8_t player) {
        int heuristicVal = 0;

        uint8_t nextPlayer;
        if (player == CELL_X) {
            nextPlayer = CELL_O;
        } else {
            nextPlayer = CELL_X;
        }
        // game over
        if (pState.isEOG()) {
            // draw
            if (pState.isDraw()) {
                heuristicVal = 0;
            }
                // X or O wins
            else if ((pState.isXWin() && currentPlayer == CELL_X) || (pState.isOWin() && currentPlayer == CELL_O)) {
                heuristicVal = NaN;
            }
                // X or O loses
            else if ((pState.isOWin() && currentPlayer == CELL_X) || (pState.isXWin() && currentPlayer == CELL_O)) {
                heuristicVal = -NaN;
            }
            return heuristicVal;
        }

        int currentPlayerMarkers = 0;
        int nextPlayerMarkers = 0;

        for (int i = 0; i < 76; i++) {
            currentPlayerMarkers = 0;
            nextPlayerMarkers = 0;
            for (int j = 0; j < 4; j++) {
                if (pState.at(winCases[i][j]) == player) {
                    currentPlayerMarkers++;
                }
                if (pState.at(winCases[i][j]) == nextPlayer) {
                    nextPlayerMarkers++;
                }
            }
            if (currentPlayerMarkers == 3) {
                heuristicVal++;
            }
            if (nextPlayerMarkers == 3) {
                heuristicVal--;
            }
        }

        if (player == currentPlayer) {
            return heuristicVal;
        } else {
            return -heuristicVal;
        }

    }


/**
 * @param gameState the current state we are analyzing
 * @param depth     the current depth we are in
 * @param alpha     the current best value achievable by A
 * @param beta      the current best value achievable by B
 * @param player    the current player
 * @return the minimax value of the state
 */
    int Player::alphaBeta(const GameState &pState, const uint8_t player, int depth, int alpha, int beta) {

        if (depth == 0) {
            return getHeuristicVal(pState, player);
        }

        vector<GameState> lNextStates;
        pState.findPossibleMoves(lNextStates);

        if (lNextStates.size() == 0) {
            return getHeuristicVal(pState, player);
        }

        int value;

        if (player == CELL_X) {
            value = -NaN;
            for (int i = 0; i < lNextStates.size(); i++) {
                value = max(value, alphaBeta(lNextStates[i], CELL_O, depth - 1, alpha, beta));
                alpha = max(alpha, value);
                if (beta <= alpha) {
                    break;
                }
            }
        } else {
            value = NaN;
            for (int i = 0; i < lNextStates.size(); i++) {
                value = min(value, alphaBeta(lNextStates[i], CELL_X, depth - 1, alpha, beta));
                beta = min(beta, value);
                if (beta <= alpha) {
                    break;
                }
            }
        }

        return value;
    }

/*namespace TICTACTOE3D*/ }

File: TTT/src/Player.java
import java.util.Vector;

public class Player {

    private static final int DEPTH = 4;
    private static final int NaN = Integer.MAX_VALUE;
    private int currentPlayer;

    /**
     * Performs a move
     *
     * @param gameState the current state of the board
     * @param deadline  time before which we must have returned
     * @return the next state the board is in after our move
     */
    public GameState play(final GameState gameState, final Deadline deadline) {

        Vector<GameState> nextStates = new Vector<GameState>();
        gameState.findPossibleMoves(nextStates);

        if (nextStates.size() == 0) {
            // Must play "pass" move if there are no other moves possible.
            return new GameState(gameState, new Move());
        }

        /**
         * Here you should write your algorithms to get the best next move, i.e.
         * the best next state. This skeleton returns a random move instead.
         */

//        Random random = new Random();
//        return nextStates.elementAt(random.nextInt(nextStates.size()));

        currentPlayer = gameState.getNextPlayer();
        int nextPlayer;
        if (currentPlayer == Constants.CELL_X)
            nextPlayer = Constants.CELL_O;
        else
            nextPlayer = Constants.CELL_X;

        int maxHeuristicVal = -NaN;
        int maxValIndex = 0;
        int heuristicVal;

        int alpha = -NaN;
        int beta = NaN;

        for (int i = 0; i < nextStates.size(); i++) {
            heuristicVal = alphaBeta(nextStates.elementAt(i), DEPTH, alpha, beta, nextPlayer);
            if (heuristicVal > maxHeuristicVal) {
                maxHeuristicVal = heuristicVal;
                maxValIndex = i;
            }
        }

        return nextStates.elementAt(maxValIndex);
    }

    /**
     * @param gameState the current state we are analyzing
     * @param depth     the current depth we are in
     * @param alpha     the current best value achievable by A
     * @param beta      the current best value achievable by B
     * @param player    the current player
     * @return the minimax value of the state
     */
    private int alphaBeta(GameState gameState, int depth, int alpha, int beta, int player) {

        Vector<GameState> nextStates = new Vector<>();
        gameState.findPossibleMoves(nextStates);

        if (depth == 0 || nextStates.size() == 0) {
            return getHeuristicVal(gameState, player);
        }

        int value;

        // player = A (CELL_X)
        if (player == Constants.CELL_X) {
            value = -NaN;
            for (int i = 0; i < nextStates.size(); i++) {
                value = Math.max(value, alphaBeta(nextStates.elementAt(i),
                        depth - 1, alpha, beta, Constants.CELL_O));
                alpha = Math.max(alpha, value);
                if (beta <= alpha)
                    break; // β prune
            }
        }

        // player = B (CELL_O)
        else {
            value = NaN;
            for (int i = 0; i < nextStates.size(); i++) {
                value = Math.min(value, alphaBeta(nextStates.elementAt(i),
                        depth - 1, alpha, beta, Constants.CELL_X));
                beta = Math.min(beta, value);
                if (beta <= alpha)
                    break; // α prune
            }
        }

        return value;
    }

    private int getHeuristicVal(GameState gameState, int player) {
        int heuristicVal;
        int rowMarks = 0;
        int colMarks = 0;
        int digMarks;

        // game over
        if (gameState.isEOG()) {
            // draw
            if (gameState.getMove().isDraw()) {
                heuristicVal = 0;
            }
            // X or O wins
            else if ((gameState.getMove().isXWin() && currentPlayer == Constants.CELL_X)
                    ||
                    (gameState.getMove().isOWin() && currentPlayer == Constants.CELL_O)) {
                heuristicVal = NaN;
            }
            // X or O loses
            else {
                heuristicVal = -NaN;
            }
            return heuristicVal;
        }

        // check rows
        for (int i = 0; i < DEPTH; i++) {
            for (int j = 0; j < DEPTH; j++) {
                if (gameState.at(i, j) == player) {
                    rowMarks++;
                    break;
                }
            }
        }

        // check columns
        for (int i = 0; i < DEPTH; i++) {
            for (int j = 0; j < DEPTH; j++) {
                if (gameState.at(j, i) == player) {
                    colMarks++;
                    break;
                }
            }
        }

        // check diagonals
        int dig = 0;
        int dig_off = 0;
        for (int i = 0; i < DEPTH; i++) {
            for (int j = 0; j < DEPTH; j++) {
                if (i == j && gameState.at(i, j) == player) {
                    dig = 1;
                }
                if ((i + j == 3) && (gameState.at(i, j) == player)) {
                    dig_off = 1;
                }
            }
        }
        digMarks = dig + dig_off;

        heuristicVal = rowMarks + colMarks + digMarks;

        if (currentPlayer == player) {
            return heuristicVal;
        } else {
            return -heuristicVal;
        }
    }

}
